### PR TITLE
implement inverse matrix calculation + implement mat4 * vec4

### PR DESCRIPTION
**Pull request by Zubnix**

### CODE REVIEW

File: src/main/java/com/hackoeur/jglm/Mat.java [hunk 1]
```diff
@@ -41,14 +41,14 @@
  * 
  * @author James Royalty
  */
-public interface Mat {
+public interface Mat<T extends Vec> {
```
Comment (jroyalty) [inline src/main/java/com/hackoeur/jglm/Mat.java]: Why did you need to parametrize the interface?

Comment (Zubnix) [inline src/main/java/com/hackoeur/jglm/Mat.java]: Because else a hidden but explicit cast happens when you get a column from a matrix. Not only do you need to manually cast it when the compiler can not work out to which vector type you're assigning, it's also extra runtime overhead (because the jvm has to check if the cast is valid).

Admitted, you don't *need* to parameterize the interface, but if you want to fix above mentioned situation, you'll have to replace all 'T' with the correct vector type directly in the matrix itself.

File: src/main/java/com/hackoeur/jglm/Mat3.java [hunk 2]
```diff
@@ -231,16 +231,16 @@ public int getNumColumns() {
 	
 	@SuppressWarnings("unchecked")
 	@Override
-	public <T extends Vec> T getColumn(final int columnIndex) {
```
Comment (Zubnix) [inline src/main/java/com/hackoeur/jglm/Mat3.java]: Here a hidden cast will happen, which you'll have to do explicitly anyway if the compiler can not work out which vector type you're assigning to.

File: src/main/java/com/hackoeur/jglm/Mat.java (hunk 1 above)
Comment (jroyalty) [inline src/main/java/com/hackoeur/jglm/Mat.java]: @Zubnix: I would prefer not to have to change the signature of the interface for this patch, breaking compatibility for others.

**Comment by jroyalty** [comment]
@Zubnix: You still interested in this PR?  I'm willing to merge it, just don't wanna cut a new version for it...

**Comment by Zubnix** [comment]
Yeah sure, I'll see if I can isolate out the invert method. Is that ok for you?

**Comment by jroyalty** [comment]
Yeah, everything is good save the genericized `Mat` interface.  I'm fine making interface changes, BTW; just gotta do a major version release for that.

**Comment by TheFakeMontyOnTheRun** [comment]
Hello! Nice code, folks! Any plans for having this Pull Request merged?

**Comment by jroyalty** [comment]
Yeah, just gotta drop the generic on `Mat`.  If you wanna do then go for it.

**Comment by TheFakeMontyOnTheRun** [comment]
Gonna try my hands at it on this weekend ;-)

**Comment by TheFakeMontyOnTheRun** [comment]
I can't change this branch (at least, I'm not sure how, with the confusing GitHub UI). Gonna cherrypick @Zubnix commits and open a new pull request. Are you guys ok with that? If not, can you suggest me another course of action?

**Comment by jroyalty** [comment]
Closing this.  Its superseded by pending changes from @TheFakeMontyOnTheRun
